### ISSUE TITLE
How can I fix Drift issues

### ISSUE THREAD

**Issue by mendelg** [MahmoudAshraf97/ctc-forced-aligner#84]
Thanks for this aligner!
It's working pretty well. However, with some words/paragraphs, there might be a Drift issue - which means, the Audio might be ahead of the text by upto ~5 seconds.

What are some ways to improve accuracy? I've already tried to play around with the `<star> `.

I'm working on Audio and Transcripts that are in the Yiddish (`yid`) language.

If it helps, I'mm attatching my code below:

```python
#!/usr/bin/env python3
import argparse
import json
import logging
import re
import sys
from pathlib import Path

import numpy as np
import torch
import librosa
from tqdm import tqdm

# --- 1. Imports for CTC Forced Aligner ---
try:
    from ctc_forced_aligner import (
        load_alignment_model, 
        generate_emissions, 
        preprocess_text,
        get_alignments, 
        get_spans, 
        postprocess_results
    )
except ImportError:
    print("Error: ctc_forced_aligner not found. Run: pip install ctc-forced-aligner")
    sys.exit(1)

logging.basicConfig(level=logging.INFO, format="%(asctime)s │ %(levelname)-7s │ %(message)s")
logger = logging.getLogger("yid_hybrid_final")

def norm_token(t: str) -> str:
    if not t: return ""
    t = re.sub(r"[\u0591-\u05C7]", "", t.strip().lower())
    t = t.replace("״", '"').replace("׳", "'")
    return re.sub(r"[^\w\u0590-\u05FF]+", "", t)

# --- 2. VAD Logic ---
def get_vad_segments(audio, sr):
    """Detects speech energy to act as a 'mask' for CTC peaks."""
    model, utils = torch.hub.load(repo_or_dir='snakers4/silero-vad', model='silero_vad', trust_repo=True)
    (get_speech_timestamps, _, _, _, _) = utils
    wav = torch.from_numpy(audio).float()
    speech_timestamps = get_speech_timestamps(wav, model, sampling_rate=sr, threshold=0.35)
    return [(ts['start'] / sr, ts['end'] / sr) for ts in speech_timestamps]

def snap_to_vad(words, vad_segments):
    """Enforces that words must exist within VAD-detected speech blocks."""
    refined = []
    for w in words:
        w_mid = (w["start"] + w["end"]) / 2
        best_seg = next((s for s in vad_segments if s[0] <= w_mid <= s[1]), None)
        
        if best_seg:
            # Anchor to physical speech boundaries
            w["start"] = max(w["start"], best_seg[0])
            w["end"] = min(w["end"], best_seg[1])
        else:
            # Snap to nearest speech if CTC drifted into silence/music
            closest = min(vad_segments, key=lambda x: min(abs(w_mid - x[0]), abs(w_mid - x[1])))
            if w_mid < closest[0]:
                w["start"], w["end"] = closest[0], closest[0] + 0.1
            else:
                w["start"], w["end"] = closest[1] - 0.1, closest[1]
        refined.append(w)
    return refined

# --- 3. Main Logic ---
def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--audio", required=True)
    ap.add_argument("--text", required=True)
    ap.add_argument("--skip", type=float, default=0.0)
    ap.add_argument("--end_time", type=float, default=None)
    ap.add_argument("--visual_offset", type=float, default=220.0)
    ap.add_argument("--merge_threshold", type=float, default=0.05, help="Fixes word gathering")
    ap.add_argument("--out", default="aligned_words.json")
    ap.add_argument("--language", default="yid")
    args = ap.parse_args()

    device = "cuda" if torch.cuda.is_available() else "cpu"
    
    # Audio Setup
    logger.info(f"Loading {args.audio} and slicing...")
    wav_np, sr = librosa.load(args.audio, sr=16000, mono=True)
    start_idx = int(args.skip * sr)
    end_idx = int(args.end_time * sr) if args.end_time else len(wav_np)
    wav_slice = wav_np[start_idx:end_idx]
    
    # VAD Pass
    logger.info("Scanning for physical speech boundaries (VAD)...")
    vad_segs = get_vad_segments(wav_slice, sr)

    # Model Setup
    dtype = torch.float16 if device == "cuda" else torch.float32
    logger.info(f"Loading model on {device}...")
    model, tokenizer = load_alignment_model(device, dtype=dtype)
    
    # Text Setup with segment-level anchoring
    tokens = [norm_token(w) for w in Path(args.text).read_text(encoding="utf-8").split() if norm_token(w)]
    transcript_str = " ".join(tokens)
    
    # PREPROCESS: Generate emissions with windowing context
    wav_tensor = torch.from_numpy(wav_slice).to(device=model.device, dtype=model.dtype)
    logger.info("Generating emissions...")
    emissions, stride = generate_emissions(model, wav_tensor, batch_size=8)
    
    # PREPROCESS TEXT: Use 'segment' frequency for star tokens to reduce internal drift
    t_star, txt_star = preprocess_text(transcript_str, romanize=True, language=args.language)
    
    # ALIGN
    segments, scores, blank_token = get_alignments(emissions, t_star, tokenizer)
    spans = get_spans(t_star, segments, blank_token)
    
    # Use merge_threshold to smooth out 'gathered' peaks
    word_ts = postprocess_results(txt_star, spans, stride, scores, merge_threshold=args.merge_threshold)

    # Refine
    raw_words = []
    for r in word_ts:
        raw_words.append({
            "text": norm_token(getattr(r, "text", r.get("text", ""))),
            "start": getattr(r, "start", r.get("start", 0)),
            "end": getattr(r, "end", r.get("end", 0))
        })
    
    logger.info("Snapping CTC data to VAD rails...")
    snapped_words = snap_to_vad(raw_words, vad_segs)

    # Final Time Mapping
    final_output = []
    for w in snapped_words:
        final_output.append({
            "text": w["text"],
            "start": round(w["start"] + args.skip + args.visual_offset, 3),
            "end": round(w["end"] + args.skip + args.visual_offset, 3)
        })

    Path(args.out).write_text(json.dumps(final_output, ensure_ascii=False, indent=2), encoding="utf-8")
    logger.info(f"Done! Saved {len(final_output)} words. Final word ends at {final_output[-1]['end']}s")

if __name__ == "__main__":
    main()
```

Thank you!

**Comment by MahmoudAshraf97** [comment]
Can you tune `generate_emissions`, especially the `window_length` and `context_length` parameters, and see if it makes a difference?

**Comment by mendelg** [comment]
Thanks for the suggestion @MahmoudAshraf97 .
It did indeed help to tune `generate_emissions`!

However, it turns out that the core issue is that the default model doesn't support Yiddish too well.
Do you know of a model that would better support of the Yiddish language?
(Or should I open a separate issue?)

**Comment by MahmoudAshraf97** [comment]
I don't know unfortunately, but any ctc asr model trained on Yiddish should be a good starting point

**Comment by moeelf** [comment]
> Thanks for the suggestion [@MahmoudAshraf97](https://github.com/MahmoudAshraf97) . It did indeed help to tune `generate_emissions`!
> 
> However, it turns out that the core issue is that the default model doesn't support Yiddish too well. Do you know of a model that would better support of the Yiddish language? (Or should I open a separate issue?)

How did you fix it? I have the same problem.

**Comment by mendelg** [comment]
@moeelf 
I was able to get better results by playing around with `generate_emissions`. But the results still weren't the best.

I have had no choice but to use shorter audio length files. Instead of a 60 min audio recording, I've split it into three: 20, 40, and 60 minutes.  and provide separate transcripts for each. 

And at the end, run a script that combines into one long audio file.

If someone has a solution, let us know!
